### ISSUE TITLE
Wrong Batch Normalization

### ISSUE THREAD

**Issue by bryant03** [Kyubyong/transformer#40]
In function normalize()
`

        with tf.variable_scope(scope, reuse=reuse):
            inputs_shape = inputs.get_shape()
            params_shape = inputs_shape[-1:]
            mean, variance = tf.nn.moments(inputs, [-1], keep_dims=True)
            print ('mean.get_shape()',mean.get_shape())
            beta= tf.Variable(tf.zeros(params_shape))
            gamma = tf.Variable(tf.ones(params_shape))
            normalized = (inputs - mean) / ( (variance + epsilon) ** (.5) )
            outputs = gamma * normalized + beta
`
but i think the second parameter of tf.nn.moments() should not be [-1], since we need to consider the batch information.
After modification the code shown as below:

`

     with tf.variable_scope(scope, reuse=reuse):
            inputs_shape = inputs.get_shape()
            params_shape = inputs_shape[-1:]
            axis = list(range(len(inputs_shape) - 1))
            mean, variance = tf.nn.moments(inputs, axis, keep_dims=True)
            print ('mean.get_shape()',mean.get_shape())
            beta= tf.Variable(tf.zeros(params_shape))
            gamma = tf.Variable(tf.ones(params_shape))
            normalized = (inputs - mean) / ( (variance + epsilon) ** (.5) )
            outputs = gamma * normalized + beta
`

**Comment by RayXu14** [comment]
Transformer use Layer Normalization rather than batch normalization. Layer Normalization need to consider the batch information. see [(https://arxiv.org/pdf/1607.06450.pdf)] at the end of page 2

**Comment by RoyJoyRo** [comment]
> Transformer use Layer Normalization rather than batch normalization. Layer Normalization need not consider the batch information. see [Layer Normalization](https://arxiv.org/pdf/1607.06450.pdf) at the end of page 2

However, I doubt that the implementation of `layer_norm` is still wrong. The paper is not describing the model clearly. I suggest referencing to the implementation in [https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/contrib/layers/python/layers/layers.py](layers.py), and change the code as `axis = list(range(1, len(inputs_shape)))`

**Comment by RayXu14** [comment]
> > Transformer use Layer Normalization rather than batch normalization. Layer Normalization need not consider the batch information. see [Layer Normalization](https://arxiv.org/pdf/1607.06450.pdf) at the end of page 2
> 
> However, I doubt that the implementation of `layer_norm` is still wrong. The paper is not describing the model clearly. I suggest referencing to the implementation in [layers.py](https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/contrib/layers/python/layers/layers.py), and change the code as `axis = list(range(1, len(inputs_shape)))`

Agree.
In this reposity, the author just use a way that is just related to last dim, not related to sequence length direction. But in layer normalization cases, generally speaking, it should be just batch irrelevant.

**Comment by pobei** [comment]
layer Normalization normalized input among the hidden states for single example, so the number of hidden state is this case should be  the depth(the number 8)?while [-1:] is 512
